### PR TITLE
feat(terrain): add Elevation Profile as a built-in plugin

### PR DESCRIPTION
**Pull request by giswqs**
## Summary

Ports the external [`geolibre-elevation-profile`](https://github.com/opengeos/geolibre-elevation-profile) marketplace plugin **in-house as a first-class built-in**, alongside Street View, Swipe, and the other bundled control plugins. It appears in the **Plugins** menu (with the usual position submenu).

Draw a line on the map to chart its **elevation profile**:
- Distance, min/max elevation, and total ascent/descent stats
- Interactive hover readout with a synced marker on the map
- Metric ⇄ imperial toggle
- CSV and SVG export (uses the native save dialog on desktop)

Elevations are sampled from the key-less **Open-Meteo** elevation API.

## What changed

- **New built-in** under `packages/plugins/src/plugins/elevation-profile/`. The control code is byte-for-byte the published plugin; only the plugin entry (`index.ts`) is rebound onto GeoLibre's built-in `GeoLibrePlugin` contract.
- Registered in `apps/geolibre-desktop/src/hooks/usePlugins.ts` and exported from `@geolibre/plugins`.
- Built-in plugins don't import CSS, so the control styles move into `apps/geolibre-desktop/src/index.css`, with a `.dark` override so the floating panel tracks the **in-app theme toggle** (not just the OS `prefers-color-scheme`), matching the repo convention.
- **State + deep links:** the drawn line, unit system, and collapsed state round-trip through the project file, and `?elevation-line=lng,lat;lng,lat` restores a shared profile on load.
- **Tests:** adds `tests/elevation-profile.test.ts` (`node:test`) covering the ported geometry, format, CSV, chart, deep-link, and Open-Meteo client modules (22 assertions).

## Verification

- `npm run build` ✅ · `eslint` ✅ · full frontend suite (2149 tests) ✅ · `pre-commit run` (incl. npm build) ✅
- **Real app (Playwright), light + dark themes:** activate from the Plugins menu → draw a line → the chart, stats grid, hover readout, and metric/imperial conversion all render correctly; the line + vertices draw on the map.

| Light | Dark |
|---|---|
| panel + orange profile chart render over the globe | dark panel surface, light text, orange chart |

## Summary by CodeRabbit

* **New Features**
  * Added an Elevation Profile plugin to the desktop app, allowing users to draw a line on the map and view elevation statistics and an interactive profile chart.
  * Added metric/imperial units, plus CSV and SVG export of the drawn profile.
  * Added deep-link support to load a pre-defined line from the app URL.
* **Bug Fixes**
  * Improved robustness for edge cases (empty/short/degenerate lines) and safer handling of async elevation fetches.
* **Tests**
  * Added comprehensive automated coverage for geometry, formatting, CSV export, deep-link parsing, and elevation fetching.

### CODE REVIEW

File: packages/plugins/src/plugins/elevation-profile/index.ts [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    return control?.getState() ?? pendingState ?? undefined;
+  },
+
+  applyProjectState(_app, state) {
+    if (!isPluginState(state)) return false;
+    pendingState = state as Partial<ElevationProfileState> & {
+      unitSystem?: UnitSystem;
+    };
+    control?.setState(pendingState);
+  },
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/index.ts:161]: Fixed in 0de99ab — applyProjectState now clears the cached line/unit/collapse to defaults when passed missing/invalid state (the New Project reset path), matching maplibre-swipe / maplibre-graticule. Good catch.

File: packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts [hunk 2]
```diff
@@ -0,0 +1,1056 @@
...
+
+    if (!this._state.collapsed) {
+      this._panel.classList.add('expanded');
+      requestAnimationFrame(() => this._updatePanelPosition());
+    }
+
+    // Restore a previously saved or deep-linked line, if any.
+    if (this._state.line && this._state.line.length >= 2) {
+      void this._profileLine(this._state.line, { fit: false });
+    }
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts:176]: Fixed in 0de99ab — onAdd now skips the re-fetch when _profilePoints.length >= 2 (control re-mounted with its profile still in memory, e.g. a reposition) and just redraws the line/vertex layers that onRemove tore down.

File: packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts [hunk 3]
```diff
@@ -0,0 +1,1056 @@
...
+    this._setStatus('');
+    this._renderProfile();
+    this._syncButtons();
+  }
+
+  // --- Map layers --------------------------------------------------------
+
+  private _ensureMapLayers(): boolean {
+    const map = this._map;
+    if (!map || !map.isStyleLoaded()) return false;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts]: Fixed in 0de99ab — _ensureMapLayers now queues a one-time map.once('idle') retry when the style isn't loaded yet (mirroring maplibre-graticule's whenStyleReady), re-running the pending draw/line render so a restored or deep-linked line still appears.

File: packages/plugins/src/plugins/elevation-profile/elevation/client.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  const longitudes = points.map((p) => p[0].toFixed(6)).join(',');
+  const url = `${ENDPOINT}?latitude=${latitudes}&longitude=${longitudes}`;
+
+  let response: Response;
+  try {
+    response = await doFetch(url);
+  } catch (error) {
+    const detail = error instanceof Error ? error.message : 'unknown error';
+    throw new ElevationFetchError(`Could not reach the elevation service: ${detail}`);
+  }
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/elevation/client.ts]: Fixed in 0de99ab — added an AbortController timeout (ELEVATION_REQUEST_TIMEOUT_MS = 15s); FetchLike now takes an optional init, and an AbortError surfaces as an ElevationFetchError('Elevation request timed out.'). Added a test for the aborted path.

File: tests/elevation-profile.test.ts [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+describe("elevation-profile Open-Meteo client", () => {
+  const stubFetch =
+    (body: unknown, ok = true, status = 200): FetchLike =>
+    () =>
+      Promise.resolve(
+        new Response(JSON.stringify(body), {
+          status: ok ? status : status,
+        }),
+      );
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/elevation-profile.test.ts]: Fixed in 0de99ab — dropped the redundant `ok` parameter from stubFetch; the Response status alone determines ok.

File: packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts [hunk 6]
```diff
@@ -0,0 +1,1056 @@
...
+
+  private _exportSvg(): void {
+    const svg = this._buildExportSvg();
+    if (!svg) return;
+    this._saveFile('elevation-profile.svg', svg, {
+      description: 'SVG image',
+      extensions: ['svg'],
+      mimeType: 'image/svg+xml',
+    });
+  }
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts:900]: Fixed in 0de99ab — both exports now pass promptName: true, and ExportFileOptions (core/types.ts) gained the promptName field so it threads through to the host's exportTextFile.

File: packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts [hunk 7]
```diff
@@ -0,0 +1,1056 @@
...
+      maxLat = Math.max(maxLat, lat);
+    }
+    this._map.fitBounds(
+      [
+        [minLng, minLat],
+        [maxLng, maxLat],
+      ],
+      { padding: 60, duration: 600, maxZoom: 14 },
+    );
+  }
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts:565]: Fixed in 0de99ab — _fitToLine now unwraps longitudes into the previous point's 360° window (matching maplibre-graticule's unwrappedLongitudeRange), so an antimeridian-crossing line fits to a tight box instead of the whole globe.

File: packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts [hunk 8]
```diff
@@ -0,0 +1,1056 @@
...
+const LINE_COLOR = '#f97316';
+const HOVER_COLOR = '#ef4444';
+
+const CHART_HEIGHT = 132;
+
+const DEFAULT_OPTIONS: Required<
+  Omit<ElevationProfileControlOptions, 'exportTextFile'>
+> = {
+  collapsed: true,
+  title: 'Elevation Profile',
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts:55]: Valid — this is a real CLAUDE.md gap. I'm deferring it to a focused follow-up rather than doing it in this review pass, because the ported control wasn't built for live label swapping: doing it properly means adding an ElevationProfileLabels host-push interface (like setGraticuleLabels / setReverseGeocodeLabels), threading it through the ~18 strings (including the dynamic status messages), wiring setElevationProfileLabels into TopToolbar's language-change effect, and adding the elevationProfile.* keys to en.json. Leaving this thread open so it isn't lost; happy to do it here instead if you'd prefer it in this PR.

File: apps/geolibre-desktop/src/hooks/usePlugins.ts [hunk 9]
```diff
@@ -142,6 +143,7 @@ manager.registerAll([
   maplibreGeoAgentPlugin,
   maplibreUsgsLidarPlugin,
   maplibreStreetViewPlugin,
+  maplibreElevationProfilePlugin,
```
[the remark this answers is not in the document]
Comment (giswqs) [reply apps/geolibre-desktop/src/hooks/usePlugins.ts:146]: Reasonable consistency point — Directions and Reverse Geocode gate coordinate lookups behind a one-time consent notice, and this sends drawn-line coordinates to Open-Meteo. I'm leaving this open for a maintainer call rather than adding it unilaterally: it's a product/UX decision (a blocking notice adds friction) and a cross-cutting change (new elevation-consent lib + useConsentGatedActions + ConsentNoticeDialogs + gating the toggle in PluginsMenu and the DesktopShell project-restore path, plus the dialog's own i18n). Note Open-Meteo elevation is public terrain data, so the exposure is the coordinates themselves — same class as reverse-geocode. Happy to implement it to match the existing pattern if you want the gate.

File: packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts [hunk 10]
```diff
@@ -0,0 +1,1098 @@
...
+    this._drawVertices = [];
+    this._profilePoints = [];
+    this._sampledCoords = [];
+    this._stats = null;
+    this._renderLineGeometry([]);
+    this._clearHover();
+    this._setStatus('');
+    this._renderProfile();
+    this._syncButtons();
+  }
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts:433]: Fixed in 1a21fcb — _clearProfile now calls _setBusy(false) so clicking Clear mid-fetch can no longer leave the draw button permanently disabled. Nice trace.

File: packages/plugins/src/plugins/elevation-profile/utils/deep-link.ts [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  return trimmed ? trimmed : null;
+}
+
+/**
+ * Encode a polyline as a compact `lng,lat;lng,lat;...` string for a share URL.
+ *
+ * @param coords - Ordered polyline vertices as `[lng, lat]`
+ * @returns The encoded string (coordinates rounded to 6 decimal places)
+ */
+export function encodeLine(coords: LngLat[]): string {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/utils/deep-link.ts:40]: Addressed in 1a21fcb — kept encodeLine (it's the tested inverse of parseLine and the canonical ?elevation-line= format) and added a doc comment noting the control only consumes the param today, so the encoder is used by tests / external callers (e.g. the Python API) and reserved for a future copy-link button. A share affordance is a reasonable follow-up.

File: packages/plugins/src/plugins/elevation-profile/index.ts [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+  applyProjectState(_app, state) {
+    if (!isPluginState(state)) {
+      // A missing/invalid state (e.g. the "New Project" reset, which calls
+      // applyProjectState(app, undefined) via restoreProjectState's
+      // resetMissingSettings) must still clear any cached line/unit/collapse,
+      // otherwise re-enabling the plugin on the new blank project would restore
+      // the previous project's profile. Mirrors maplibre-swipe /
+      // maplibre-graticule, whose normalizers reset to defaults on undefined.
+      const cleared: ElevationProfileState = {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/index.ts:148]: Leaving as-is by design. applyProjectState runs once per project load / New Project reset (not a hot path), and setState's reference check on `line` already prevents a redundant Open-Meteo fetch for the null-vs-null reset case; worst case is one extra _renderProfile() on load. Adding a full state-equality helper for that mirrors graticule but buys negligible savings here for added surface, so I'm keeping the simpler form.

File: packages/plugins/src/plugins/elevation-profile/index.ts [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+function isLngLatArray(value: unknown): value is LngLat[] {
+  return (
+    Array.isArray(value) &&
+    value.every(
+      (pair) =>
+        Array.isArray(pair) &&
+        pair.length === 2 &&
+        typeof pair[0] === "number" &&
+        typeof pair[1] === "number",
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/index.ts:63]: Reasonable hardening. Since bad coordinates from a hand-edited .geolibre.json only cause a failed Open-Meteo fetch (surfaced as an error message), not a crash or security issue, I'm treating the finite/range tightening + guard tests as a non-blocking follow-up rather than holding the merge. Tracked.

[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/index.ts:63]: Same as the thread above — noted as an optional hardening follow-up (finite/range validation to match parseLine, plus isPluginState/isLngLatArray tests).

File: packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts [hunk 14]
```diff
@@ -0,0 +1,1104 @@
...
+      geometry.padding.left - 4,
+      geometry.yScale(geometry.maxElevation) + 3,
+      'end',
+    );
+    const minLabel = this._axisLabel(
+      formatElevation(geometry.minElevation, system),
+      geometry.padding.left - 4,
+      geometry.yScale(geometry.minElevation),
+      'end',
+    );
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts:812]: Low-priority polish; a perfectly flat profile is rare. Noted for a follow-up (skip/nudge the min label when the elevation range is ~0).

File: packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts [hunk 15]
```diff
@@ -0,0 +1,1104 @@
...
+
+  private _dedupeTailVertex(): void {
+    const n = this._drawVertices.length;
+    if (n < 2) return;
+    const a = this._drawVertices[n - 1];
+    const b = this._drawVertices[n - 2];
+    if (Math.abs(a[0] - b[0]) < 1e-7 && Math.abs(a[1] - b[1]) < 1e-7) {
+      this._drawVertices.pop();
+    }
+  }
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/plugins/src/plugins/elevation-profile/core/ElevationProfileControl.ts:352]: Low-confidence edge case inherited from the upstream plugin. A pixel-based dedup (project the last two vertices and compare screen distance) would be the cleaner fix than a fixed-degree epsilon. Noted, not blocking.
